src/utils/format_open_telemetry/format_open_telemetry.cc

```diff
@@ -0,0 +1,168 @@
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ **/
+
+extern "C" {
+#include "collectd.h"
+#include "metric.h"
+}
+
+#include "utils/format_open_telemetry/format_open_telemetry.h"
+
+#include "opentelemetry/proto/collector/metrics/v1/metrics_service.pb.h"
+#include "opentelemetry/proto/common/v1/common.pb.h"
+#include "opentelemetry/proto/metrics/v1/metrics.pb.h"
+#include "opentelemetry/proto/resource/v1/resource.pb.h"
+
+using opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest;
+using opentelemetry::proto::common::v1::AnyValue;
+using opentelemetry::proto::common::v1::InstrumentationScope;
+using opentelemetry::proto::common::v1::KeyValue;
+using opentelemetry::proto::metrics::v1::AGGREGATION_TEMPORALITY_CUMULATIVE;
+using opentelemetry::proto::metrics::v1::Gauge;
+using opentelemetry::proto::metrics::v1::Metric;
+using opentelemetry::proto::metrics::v1::NumberDataPoint;
+using opentelemetry::proto::metrics::v1::ResourceMetrics;
+using opentelemetry::proto::metrics::v1::ScopeMetrics;
+using opentelemetry::proto::metrics::v1::Sum;
+using opentelemetry::proto::resource::v1::Resource;
+
+static void metric_to_number_data_point(NumberDataPoint *dp,
+                                        metric_t const *m) {
+  for (size_t i = 0; i < m->label.num; i++) {
+    label_pair_t *l = m->label.ptr + i;
+
+    KeyValue *kv = dp->add_attributes();
+    kv->set_key(l->name);
+    AnyValue *v = kv->mutable_value();
+    v->set_string_value(l->value);
+  }
+
+  dp->set_time_unix_nano(CDTIME_T_TO_NS(m->time));
+  // TODO(): also set "start time". We may need to use the cache to determine
+  // when we've seen a metric for the first time.
+
+  switch (m->family->type) {
+  case METRIC_TYPE_COUNTER:
+    dp->set_as_int(m->value.derive);
+    break;
+  case METRIC_TYPE_GAUGE:
+    dp->set_as_double(m->value.gauge);
+    break;
+  case METRIC_TYPE_UNTYPED:
+    // TODO
+    assert(0);
```

Comment:
Maybe `FATAL()`?

Reply:
Ugh, yeah, that's still in there …

I think `METRIC_TYPE_UNTYPED` as is exists today has problems:

* It is not zero. So it cannot used to detect broken plugins that fail to initialize the structure properly.
* It makes it appear as if untyped metrics would somehow be legal. Maybe the renaming it to "uninitialized" would be better.

That is for another PR though. For now, I'll make sure to skip metric families entirely.

Reply:
Done